src/openms/source/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.cpp

```diff
@@ -0,0 +1,342 @@
+// --------------------------------------------------------------------------
+// $Maintainer: Ahmed Khalil $
+// $Authors: Ahmed Khalil $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.h>
+
+namespace OpenMS
+{
+  IsotopeLabelingMDVs::IsotopeLabelingMDVs()
+  {
+  }
+
+  IsotopeLabelingMDVs::~IsotopeLabelingMDVs()
+  {
+  }
+  
+  void IsotopeLabelingMDVs::isotopicCorrection(
+    Feature& normalized_feature,
+    Feature& corrected_feature,
+    std::vector<std::vector<double>> correction_matrix)
+  {
+    // MDV_corrected = correction_matrix_inversed * MDV_observed (normalized_features)
+    
+    uint16_t correction_matrix_n = correction_matrix.size();
+    std::vector<std::vector<double>> correction_matrix_inversed(correction_matrix_n, std::vector<double>(correction_matrix_n,0));
+    
+    // 1- correction matrix inversion
+    inverseMatrix_(correction_matrix, correction_matrix_inversed);
+    
+    // 2- element-wise expansion with MDV_observed
+    std::vector<Feature> normalized_feature_subordinates = normalized_feature.getSubordinates();
+    std::vector<double> MDV_observed;
+    for (auto it = normalized_feature_subordinates.begin(); it != normalized_feature_subordinates.end(); it++)
+    {
+      MDV_observed.push_back(it->getMetaValue("peak_apex_int"));
+    }
+    
+    double corrected_value;
+    corrected_feature = normalized_feature;
+    for (size_t i = 0; i < correction_matrix_inversed.size(); ++i) {
+      corrected_value = 0.0;
+      for (size_t j = 0; j < correction_matrix_inversed[0].size(); ++j) {
+        corrected_value += correction_matrix_inversed[i][j] * MDV_observed[j];
+      }
+      corrected_feature.getSubordinates().at(i).setIntensity(corrected_value);
+    }
+  }
+
+  void IsotopeLabelingMDVs::isotopicCorrections(FeatureMap& normalized_featureMap, FeatureMap& corrected_featureMap, std::vector<std::vector<double>> correction_matrix)
+  {
+    for (FeatureMap::Iterator feature_it = normalized_featureMap.begin(); feature_it != normalized_featureMap.end(); feature_it++){
+      Feature corrected_feature;
+      isotopicCorrection( *feature_it, corrected_feature, correction_matrix);
+      corrected_featureMap.push_back(corrected_feature);
+    }
+  }
+
+  void IsotopeLabelingMDVs::calculateIsotopicPurity(
+    Feature& normalized_featuremap,
+    Feature& featuremap_with_isotopic_purity,
+    std::vector<double>& experiment_data,
+    std::string& isotopic_purity_name)
+  {
+    featuremap_with_isotopic_purity = normalized_featuremap;
+    
+    double experiment_data_peak = 0.0;
+    if ( !experiment_data.empty() )
+    {
+      std::vector<double>::iterator max_it  = std::max_element(experiment_data.begin(), experiment_data.end());
+      uint64_t experiment_data_peak_idx     = std::distance(experiment_data.begin(), max_it);
+      experiment_data_peak                  = experiment_data[experiment_data_peak_idx];
```

Comment:
please remove additional white spaces (onle one space before and after equal sign)